crates/typst-library/src/compute/data.rs

```diff
@@ -149,22 +157,51 @@ impl csv {
         #[named]
         #[default]
         delimiter: Delimiter,
+        /// Whether to use the header row in the CSV file to generate labels the fields in the columns.
+        /// Defaults to `{false}`.
+        #[named]
+        #[default(false)]
+        use_headers: bool,
     ) -> SourceResult<Array> {
         let Spanned { v: data, span } = data;
         let mut builder = ::csv::ReaderBuilder::new();
-        builder.has_headers(false);
+        builder.has_headers(use_headers);
         builder.delimiter(delimiter.0 as u8);
         let mut reader = builder.from_reader(data.as_slice());
+        let mut headers: Option<::csv::StringRecord> = None;
+
+        let mut line_offset: usize = 1; // Counting lines from 1
+
+        if use_headers {
+            line_offset = 2; // Counting lines from 2 (1 is header)
+            let headers_result = reader.headers();
+            headers = Some(
+                headers_result
+                    .map_err(|err| format_csv_error(err, 1))
+                    .at(span)?
+                    .clone(),
+            );
+        }
+
         let mut array = Array::new();
 
         for (line, result) in reader.records().enumerate() {
             // Original solution use line from error, but that is incorrect with
             // `has_headers` set to `false`. See issue:
             // https://github.com/BurntSushi/rust-csv/issues/184
-            let line = line + 1; // Counting lines from 1
+            let line = line + line_offset;
             let row = result.map_err(|err| format_csv_error(err, line)).at(span)?;
-            let sub = row.into_iter().map(|field| field.into_value()).collect();
-            array.push(Value::Array(sub))
+            if let Some(headers) = headers.clone() {
+                let mut dict = Dict::new();
+                for (header_field, field) in headers.into_iter().zip(row.into_iter()) {
+                    let value = field.into_value();
+                    dict.insert(header_field.into(), value)
+                }
+                array.push(dict.into_value())
```

Comment:
I think rather than cloing `headers` here and using `into_iter` we could just iterate over it without cloning.